src/modules/wlr/workspace_manager.cpp

```diff
@@ -0,0 +1,323 @@
+  if (!id.empty()) {
+    box_.get_style_context()->add_class(id);
+  }
+  event_box_.add(box_);
+
+  add_registry_listener(this);
+  if (!workspace_manager_) {
+    return;
+  }
+}
+
+auto WorkspaceManager::register_manager(wl_registry *registry, uint32_t name, uint32_t version)
+    -> void {
+  if (workspace_manager_) {
+    spdlog::warn("Register workspace manager again although already registered!");
+    return;
+  }
+  if (version != 1) {
+    spdlog::warn("Using different workspace manager protocol version: {}", version);
+  }
+  workspace_manager_ = workspace_manager_bind(registry, name, version, this);
+}
+
+auto WorkspaceManager::handle_workspace_group_create(
+    zwlr_workspace_group_handle_v1 *workspace_group_handle) -> void {
+  auto new_id = ++group_global_id;
+  groups_.push_back(
+      std::make_unique<WorkspaceGroup>(bar_, box_, config_, *this, workspace_group_handle, new_id));
+  spdlog::debug("Workspace group {} created", new_id);
+}
+
+auto WorkspaceManager::handle_finished() -> void {
+  zwlr_workspace_manager_v1_destroy(workspace_manager_);
+  workspace_manager_ = nullptr;
+}
+
+auto WorkspaceManager::handle_done() -> void {
+  for (auto &group : groups_) {
+    group->handle_done();
+  }
+  dp.emit();
+}
+
+auto WorkspaceManager::update() -> void {
+  for (auto &group : groups_) {
+    group->update();
+  }
+  AModule::update();
+}
+
+WorkspaceManager::~WorkspaceManager() {
+  if (!workspace_manager_) {
+    return;
+  }
+
+  zwlr_workspace_manager_v1_destroy(workspace_manager_);
+  workspace_manager_ = nullptr;
+}
+auto WorkspaceManager::remove_workspace_group(uint32_t id) -> void {
+  auto it = std::find_if(groups_.begin(),
+                         groups_.end(),
+                         [id](const std::unique_ptr<WorkspaceGroup> &g) { return g->id() == id; });
+
+  if (it == groups_.end()) {
+    spdlog::warn("Can't find group with id {}", id);
+    return;
+  }
+
+  groups_.erase(it);
+}
+auto WorkspaceManager::commit() -> void { zwlr_workspace_manager_v1_commit(workspace_manager_); }
+
+WorkspaceGroup::WorkspaceGroup(const Bar &bar, Gtk::Box &box, const Json::Value &config,
+                               WorkspaceManager &              manager,
+                               zwlr_workspace_group_handle_v1 *workspace_group_handle, uint32_t id)
+    : bar_(bar),
+      box_(box),
+      config_(config),
+      workspace_manager_(manager),
+      workspace_group_handle_(workspace_group_handle),
+      id_(id) {
+  add_workspace_group_listener(workspace_group_handle, this);
+  auto config_sort_by_name = config_["sort-by-name"];
+  if (config_sort_by_name.isBool()) {
+    sort_by_name = config_sort_by_name.asBool();
+  }
+
+  auto config_sort_by_coordinates = config_["sort-by-coordinates"];
+  if (config_sort_by_coordinates.isBool()) {
+    sort_by_coordinates = config_sort_by_coordinates.asBool();
+  }
+}
+auto WorkspaceGroup::add_button(Gtk::Button &button) -> void {
+  box_.pack_start(button, false, false);
+}
```

Comment:
You only add buttons to the box. But what about removing buttons again when a workspace is removed. This might happen when an output is disconnected.

Reply:
Done.